Tests/ApolloTests/LinkTests.swift

```diff
@@ -0,0 +1,433 @@
+      expectation.fulfill()
+    }
+    
+    waitForExpectations(timeout: 0.1)
+  }
+  
+  func testLinkConstructionFromNonTerminatingLinks() {
+    let terminatingLink = link(from: [
+      SetContextLink.incrementCounter(by: 1),
+      SetContextLink.incrementCounter(by: 2),
+      SetContextLink.incrementCounter(by: 3),
+    ]).concat(SyncLink())
+    
+    let expectation = self.expectation(description: "link executed")
+    
+    terminatingLink.execute(operation: query).andThen { response in
+      XCTAssertEqual(response.body["counter"] as? Int, 6)
+      expectation.fulfill()
+    }
+    
+    waitForExpectations(timeout: 0.1)
+  }
+  
+  func testLinkConstructionFromNonTerminatingLinksAndTerminatingLink() {
+    let terminatingLink = link(from: [
+      SetContextLink.incrementCounter(by: 1),
+      SetContextLink.incrementCounter(by: 2),
+      SetContextLink.incrementCounter(by: 3),
+    ], terminating: SyncLink())
+    
+    let expectation = self.expectation(description: "link executed")
+    
+    terminatingLink.execute(operation: query).andThen { response in
+      XCTAssertEqual(response.body["counter"] as? Int, 6)
+      expectation.fulfill()
+    }
+    
+    waitForExpectations(timeout: 0.1)
+  }
+  
+  func testTerminatingLinkConstructionFromNonTerminatingLinks() {
+    let terminatingLink: TerminatingLink = link(from: [
+      SetContextLink.incrementCounter(by: 1),
+      SetContextLink.incrementCounter(by: 2),
+      SetContextLink.incrementCounter(by: 3),
+      EarlyTerminatingLink()
+    ])
+    
+    let expectation = self.expectation(description: "link executed")
+    
+    terminatingLink.execute(operation: query).andThen { response in
+      XCTAssertEqual(response.body["counter"] as? Int, 6)
+      expectation.fulfill()
+    }
+    
+    waitForExpectations(timeout: 0.1)
+  }
+  
+  func testConcatLink() {
+    let link = SetContextLink.incrementCounter(by: 1).concat(SyncLink())
+    
+    let expectation = self.expectation(description: "link executed")
+    
+    link.execute(operation: query).andThen { response in
+      XCTAssertEqual(response.body["counter"] as? Int, 1)
+      expectation.fulfill()
+    }
+    
+    waitForExpectations(timeout: 0.1)
+  }
+  
+  func testConcatMultipleLinks() {
+    let link = SetContextLink.incrementCounter(by: 1).concat(SetContextLink.incrementCounter(by: 2)).concat(SyncLink())
+    
+    let expectation = self.expectation(description: "link executed")
+    
+    link.execute(operation: query).andThen { response in
+      XCTAssertEqual(response.body["counter"] as? Int, 3)
+      expectation.fulfill()
+    }
+    
+    waitForExpectations(timeout: 0.1)
+  }
+  
+  func testSplitNonTerminatingLinkFirst() {
+    let queryLink = SetContextLink.incrementCounter(by: 1)
+    let mutationLink = SetContextLink.incrementCounter(by: -1)
+    
+    let splitLink = passthroughLink.split(first: queryLink, second: mutationLink) { op, _  in
+      type(of: op).rootCacheKey == "QUERY_ROOT"
```

Comment:
We shouldn't rely on `rootCacheKey` to get the operation type (and probably not expose it). Instead, maybe we could have a `GraphQLOperationType` enum and `operationType` property.

Reply:
I would also prefer a way to avoid having to go through `type(of: op)`, making the properties non-static. (They can still be static in the generated code, but maybe we can define instance properties that forward to the static ones in a protocol extension.)